thrust/thrust/detail/type_traits.h

```diff
@@ -47,361 +47,28 @@ using integral_constant = ::cuda::std::integral_constant<T, v>;
-{};
-template <>
-struct is_integral<const unsigned int> : public true_type
-{};
-template <>
-struct is_integral<const long> : public true_type
-{};
-template <>
-struct is_integral<const unsigned long> : public true_type
-{};
-template <>
-struct is_integral<const long long> : public true_type
-{};
-template <>
-struct is_integral<const unsigned long long> : public true_type
-{};
-
-template <typename T>
-struct is_floating_point : public false_type
-{};
-template <>
-struct is_floating_point<float> : public true_type
-{};
-template <>
-struct is_floating_point<double> : public true_type
-{};
-template <>
-struct is_floating_point<long double> : public true_type
-{};
-
-template <typename T>
-struct is_arithmetic : public is_integral<T>
-{};
-template <>
-struct is_arithmetic<float> : public true_type
-{};
-template <>
-struct is_arithmetic<double> : public true_type
-{};
-template <>
-struct is_arithmetic<const float> : public true_type
-{};
-template <>
-struct is_arithmetic<const double> : public true_type
-{};
-
-template <typename T>
-struct is_pointer : public false_type
-{};
-template <typename T>
-struct is_pointer<T*> : public true_type
-{};
-
-template <typename T>
-struct is_device_ptr : public false_type
-{};
-
-template <typename T>
-struct is_void : public false_type
-{};
-template <>
-struct is_void<void> : public true_type
-{};
-template <>
-struct is_void<const void> : public true_type
-{};
-
-template <typename T>
-struct is_non_bool_integral : public is_integral<T>
+struct is_non_bool_integral : public ::cuda::std::is_integral<T>
 {};
 template <>
 struct is_non_bool_integral<bool> : public false_type
 {};
 
 template <typename T>
-struct is_non_bool_arithmetic : public is_arithmetic<T>
+struct is_non_bool_arithmetic : public ::cuda::std::is_arithmetic<T>
 {};
 template <>
 struct is_non_bool_arithmetic<bool> : public false_type
 {};
 
-template <typename T>
-struct is_pod
-    : public integral_constant<bool,
-                               is_void<T>::value || is_pointer<T>::value
-                                 || is_arithmetic<T>::value
-#if defined(_CCCL_COMPILER_MSVC) || defined(_CCCL_COMPILER_CLANG)
-                                 // use intrinsic type traits
-                                 || __is_pod(T)
-#elif defined(_CCCL_COMPILER_GCC)
-// only use the intrinsic for >= 4.3
-#  if (__GNUC__ * 100 + __GNUC_MINOR__ >= 403)
-                                 || __is_pod(T)
-#  endif // GCC VERSION
-#endif // THRUST_HOST_COMPILER
-                               >
-{};
-
-template <typename T>
-struct has_trivial_constructor
-    : public integral_constant<bool, is_pod<T>::value || ::cuda::std::is_trivially_constructible<T>::value>
-{};
-
-template <typename T>
-struct has_trivial_copy_constructor
-    : public integral_constant<bool, is_pod<T>::value || ::cuda::std::is_trivially_copyable<T>::value>
-{};
-
-template <typename T>
-struct has_trivial_destructor : public is_pod<T>
```

Comment:
praise: this is incredible. We used to call an extra kernel destructing all the elemenets of a device vector for most of the types that thrust traits haven't recognized as pod! With this PR, we do not destruct objects unless there's a non-trivial destructor. This essentially means we save a kernel call per each device vector, leading to 30% reduction in overhead of using device vector.